src/Renci.SshNet/SshCommand.cs

```diff
@@ -294,6 +294,9 @@ public string EndExecute(IAsyncResult asyncResult)
                             _channel.Close();
                         }
 
+                        this._session.Disconnected -= Session_Disconnected;
+                        this._session.ErrorOccured -= Session_ErrorOccured;

```

Comment:
Problem is that you’re not ever disposing the SshCommand.

Try modifying the code to this:

 

using Renci.SshNet;

 

namespace ConsoleApplication1

{

    class Program

    {

        static void Main(string[] args)

        {

            SshClient connection = new SshClient("127.0.0.1", "user", "password");

            connection.Connect();

            if (connection.IsConnected)

            {

                while (true)

                {

                    var cmd = connection.RunCommand("pwd");

                    cmd.Dispose();

                }

            }

        }

    }

}

 

We could definitely move event subscribe and unsubscribe from ctor / dispose to BeginExecute / EndExecute.

We’d also need to add this cleanup to CancelAsync().

 

If however, we ever want to support multiple async executes in parallel using the same SshCommand, then we have to keep the subscribe / unsubscribe as it is right now.

What we could do in that case, is to not only unsubscribe when Dispose(bool) is invoked from the Dispose() method – as part of IDisposable - but also when it’s invoked from the finalizer (with Dispose(false)). That would also fix the leak.

 

What do you think ?

 

Regards,

Gert

 

From: GenericHero [mailto:notifications@github.com] 
Sent: vrijdag 29 april 2016 23:15
To: sshnet/SSH.NET <SSH.NET@noreply.github.com>
Cc: Gert Driesen <gert.driesen@telenet.be>; Comment <comment@noreply.github.com>
Subject: Re: [sshnet/SSH.NET] Fix memory leaks in SshCommand.cs (#7)

 

In src/Renci.SshNet/SshCommand.cs <https://github.com/sshnet/SSH.NET/pull/7#discussion_r61645771> :

> @@ -294,6 +294,9 @@ public string EndExecute(IAsyncResult asyncResult)
>                              _channel.Close();
>                          }
>  
> +                        this._session.Disconnected -= Session_Disconnected;
> +                        this._session.ErrorOccured -= Session_ErrorOccured;

Something as simple as..

namespace ConsoleApplication1
{
class Program
{
static void Main(string[] args)
{
SshClient connection = new SshClient("127.0.0.1", "user", "password");
connection.Connect();
if (connection.IsConnected)
{
while (true)
{
connection.RunCommand("pwd"); //leaks here
}
}
}
}
}

will accumulate EventHandlers that are from "this._session" that keep SshCommands alive. Note that it's hard to see the memory usage from task manager from this program. My use case uses 10+ threads polling a lot and over the course of a few hours the memory usage climbs noticeably. If you use a program called ".Net Memory Profiler" it will tell you this information very neatly. It has a free trial for a few days, using this is how I discovered this particular issue.

—
You are receiving this because you commented.
Reply to this email directly or  <https://github.com/sshnet/SSH.NET/pull/7/files/6b1739d20c3acf7f89bb1380915aeccb67d18552#r61645771> view it on GitHub  <https://github.com/notifications/beacon/AATGHxJ1snt7794BMRlcZYIybUez6BHWks5p8nTGgaJpZM4ITDsp.gif>